lib/pg/instrumentation.rb

```diff
@@ -0,0 +1,158 @@
+require 'pg/instrumentation/version'
+require 'opentracing'
+
+module PG
+  module Instrumentation
+    class << self
+
+      attr_accessor :tracer
+      attr_accessor :tags
+
+      ::PG::Instrumentation.tags = {
+        'span.kind' => 'client',
+        'component' => 'pg',
+        'db.type' => 'pg'
+      }
+
+      def instrument(tracer: OpenTracing.global_tracer)
+        begin
+          require 'pg'
+        rescue LoadError
+          return
+        end
+
+        @tracer = tracer
+
+        patch_methods unless @instrumented
+        @instrumented = true
+      end
+
+      def patch_methods
+        ::PG::Connection.class_eval do
+
+          alias_method :initialize_original, :initialize
+          alias_method :exec_original, :exec
+          alias_method :exec_params_original, :exec_params
+          alias_method :prepare_original, :prepare
+          alias_method :exec_prepared_original, :exec_prepared
+
+          def initialize(*args)
+            return initialize_original(*args) if args.empty?
+
+            hash_arg = args.first.is_a?( Hash ) ? args.first : {}
+            db_name = hash_arg.fetch(:dbname, nil)
+            db_user = hash_arg.fetch(:user, nil)
+            host = hash_arg.fetch(:host, nil)
+            port = hash_arg.fetch(:port, nil)
+
+            ::PG::Instrumentation.tags['db.instance'] = db_name if db_name
+            ::PG::Instrumentation.tags['db.user'] = db_user if db_user
+            ::PG::Instrumentation.tags['peer.hostname'] = host if host
+            ::PG::Instrumentation.tags['peer.port'] = port if port
+            ::PG::Instrumentation.tags['peer.address'] = "pg://#{host}:#{port}" if host && port
+
+            operation_name = 'pg.initialize'
+
+            scope = ::PG::Instrumentation.tracer.start_active_span(operation_name, tags: ::PG::Instrumentation.tags)
+
+            initialize_original(*args)
+
+          rescue => e
+            log_error(scope.span, e) unless args.empty?
```

Comment:
Why the condition?  Wouldn't a nil check in log_error be more robust over time?  Also possible that scope would be nil?

Reply:
Followed the pattern here:
https://github.com/ged/ruby-pg/blob/a50f238f6b5627a38df81ccb5da1dbb120f270e3/lib/pg/connection.rb#L37

`self::parse_connect_args( *args )` is executed after initialize and, afaict, parses the arguments from the initializer.  However, these arguments would either be arrays, strings or hashes (https://www.rubydoc.info/gems/pg/PG/Connection#initialize-instance_method), for which `.empty?` is defined).

Based on: 
https://github.com/signalfx/ruby-pg-instrumentation/blob/a80d62c3a572760aa97e93dee8b1e0600cb91cfe/lib/pg/instrumentation.rb#L40

The scope will be nil due to the return statement when `args.empty?` evaluates to true. So`scope.span` would result in an error. Hence the need for the check.

Reply:
You are using `log_error(scope.span, e)` throughout, and scope could be nil.